stan/math/opencl/rev/copy.hpp

```diff
@@ -26,19 +94,17 @@ namespace math {
  * @param src source Eigen matrix
  * @return matrix_cl with a copy of the data in the source matrix
  */
-template <typename T, int R, int C, require_var_t<T>* = nullptr>
-inline matrix_cl<T> to_matrix_cl(const Eigen::Matrix<T, R, C>& src) try {
-  matrix_cl<T> dst(src.rows(), src.cols());
-  if (src.size() == 0) {
-    return dst;
-  }
-  dst.val() = to_matrix_cl(src.val().eval());
-  dst.adj() = to_matrix_cl(src.adj().eval());
-  return dst;
-} catch (const cl::Error& e) {
-  check_opencl_error("copy var Eigen->(OpenCL)", e);
-  matrix_cl<T> dst(src.rows(), src.cols());
-  return dst;
+template <typename T, require_eigen_vt<is_var, T>* = nullptr>
+inline var_value<matrix_cl<value_type_t<value_type_t<T>>>> to_matrix_cl(
+    T& src) {
+  // the matrix can go out of scope before chain() is called. So we store a map
+  // to the data
+  var* src_array
+      = ChainableStack::instance_->memalloc_.alloc_array<var>(src.size());
+  Eigen::Map<plain_type_t<T>> src_stacked(src_array, src.rows(), src.cols());
+  src_stacked = src;
+  return new internal::op_copy_to_cl_vari<decltype(src_stacked.adj())>(
+      src_stacked.val(), src_stacked.adj());
```

Comment:
Is a pointer to a `vari_value` inherited type implicitly convertible to a `var_value`? Or should this be like below

```suggestion
  return {new internal::op_copy_to_cl_vari<decltype(src_stacked.adj())>(
      src_stacked.val(), src_stacked.adj())};
```

Reply:
Yep, it is implicitly convertible. We have an implicit constructor.